Cloud OIDC base guide and Oracle Cloud guide

burtbeckwith opened the pull request.

burtbeckwith committed:
Cloud OIDC base guide and Oracle Cloud guide

burtbeckwith commented:
I wasn't able to generate a native image so I commented that out of the Asciidoc and reported micronaut-projects/micronaut-core#7129

@sdelamo I can share valid values for OAUTH_CLIENT_ID, OAUTH_CLIENT_SECRET, and OAUTH_ISSUER for testing

burtbeckwith commented:
Guide PDF: [oidc_oracle.pdf](https://github.com/micronaut-projects/micronaut-guides/files/8180218/oidc_oracle.pdf)

alvarosanchez commented on src/docs/common/callouts/callout-oauth2-client-name.adoc:
```diff
@@ -1 +1 @@
-The provider identifier should match the last part of the URL you entered as a redirect URL `/oauth/callback/{0}`
\ No newline at end of file
+The provider identifier must match the last part of the URL you entered as a redirect URL `/oauth/callback/{0}`
```
```suggestion
The provider identifier must match the last part of the URL you entered as a redirect URL: `/oauth/callback/{0}`
```

graemerocher committed:
Update src/docs/common/callouts/callout-oauth2-client-name.adoc

Co-authored-by: Álvaro Sánchez-Mariscal <alvaro.sanchez-mariscal@oracle.com>

burtbeckwith committed:
move Thymeleaf template to base guide

burtbeckwith commented:
@sdelamo could you get this into the proofreading queue?

sdelamo commented:
> @sdelamo could you get this into the proofreading queue?

@burtbeckwith  I will get on this now. sorry for taking this long

sdelamo committed:
Merge branch 'master' into oidc_oracle

sdelamo committed:
remove workflow files

sdelamo commented on guides/micronaut-cloud-oidc-base/end.adoc:
```diff
@@ -0,0 +1,14 @@
+// TODO get native image working
```
should we remove these TODO statements?

sdelamo commented on guides/micronaut-cloud-oidc-base/java/src/main/java/example/micronaut/HomeController.java:
```diff
@@ -0,0 +1,30 @@
+    @Secured(IS_AUTHENTICATED) // <5>
+    @Get("/secure") // <6>
+    public Map<String, Object> secured() {
+        return CollectionUtils.mapOf("secured", true); // <7>
```
can we not use `Collections.singletonMap`?

sdelamo commented on guides/micronaut-cloud-oidc-base/kotlin/src/main/kotlin/example/micronaut/HomeController.kt:
```diff
@@ -0,0 +1,21 @@
+
+    @Secured(IS_AUTHENTICATED) // <5>
+    @Get("/secure") // <6>
+    fun secured(): Map<String, Any> = mutableMapOf("secured" to true) // <7>
```
you should be able to use `mapOf`. We fixed it in https://github.com/micronaut-projects/micronaut-views/pull/353

sdelamo commented on guides/micronaut-cloud-oidc-base/metadata.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "publish": false,
+  "authors": ["Burt Beckwith"],
```
I think you should remove yourself as author from the base guide. I think base guides should not have authors.

sdelamo commented on guides/micronaut-cloud-oidc-base/src/main/resources/views/home.html:
```diff
@@ -0,0 +1,27 @@
+<body>
+<h1>Micronaut - OpenID Connect</h1>
+
+<h2 th:if="${security}">Username: <span th:text="${security.attributes.get('user_displayname')}"></span></h2>
```
why `user_displayname`? . That it is not a standard OpenID Connect claims. should not we use a standard claim such as name when you request the scope `profile`?

sdelamo commented on guides/micronaut-cloud-oidc-oracle/metadata.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "base": "micronaut-cloud-oidc-base",
+  "title": "Authenticate Micronaut Application users with OpenID Connect and Oracle IDCS",
```
@burtbeckwith add yourself as an author

sdelamo commented on guides/micronaut-cloud-oidc-base/metadata.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "publish": false,
+  "authors": ["Burt Beckwith"],
+  "tags": ["cloud", "security", "oauth2", "oidc"],
```
probably `openid-connect` as a tag as well. not sure if people identify `oidc` as an acronym.

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+Click "Next" to go to step 2.
+
+Click "Configure this application as a client now", and in the `Allowed Grant Types` section check `Client Credentials` and `Authorization Code`.
```
why `Client Credentials` you don't need it for this guide.

kroegerk replied:
```suggestion
Click "Configure this application as a client now", and in the `Allowed Grant Types` section, check `Client Credentials` and `Authorization Code`.
```

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+resource:application.yml[]
+
+<1> Set `micronaut.security.authentication` as `idtoken`. The `idtoken` provided by your OAuth 2.0 application when the  Authorization code flow ends (a signed JWT) will be saved in a cookie.
```
extract this to a `callout` snippet.

kroegerk replied:
```suggestion
<1> Set `micronaut.security.authentication` as `idtoken`. The `idtoken` provided by your OAuth 2.0 application when the Authorization code flow ends (a signed JWT) will be saved in a cookie.
```

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+<1> Set `micronaut.security.authentication` as `idtoken`. The `idtoken` provided by your OAuth 2.0 application when the  Authorization code flow ends (a signed JWT) will be saved in a cookie.
+callout:oauth2-client-name[number=2,arg0=oci]
+<3> `client-id`, `client-secret`, and the `issuer` URL will be set as environment variables
```
extract this as a `callout` snippet.

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+<1> Set `micronaut.security.authentication` as `idtoken`. The `idtoken` provided by your OAuth 2.0 application when the  Authorization code flow ends (a signed JWT) will be saved in a cookie.
+callout:oauth2-client-name[number=2,arg0=oci]
+<3> `client-id`, `client-secret`, and the `issuer` URL will be set as environment variables
+<4> Accept GET requests to the `/logout` endpoint
```
extract this as a callout snippet.

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+callout:get[number=6,arg0=secured,arg1=/secure]
+<7> This method simply returns a model map that will be rendered   as JSON since there's no `@View` annotation
+
+=== Thymeleaf view
```
should this be in a common snippet instead of the oracle guide?

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+== Running the Application
+
+Before starting the application we need to create some environment variables.
```
should this be in a common snippet.

kroegerk replied:
```suggestion
Before starting the application, we need to create some environment variables.
```

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+common:runapp-instructions.adoc[]
+
+Open http://localhost:8080/secure in a browser and you'll be redirected to the home page because the controller method is annotated with `@Secured(IS_AUTHENTICATED)` and we're not yet authenticated:
```
should this be in a common snippet in base?

kroegerk replied:
```suggestion
Open http://localhost:8080/secure in a browser, and you'll be redirected to the home page because the controller method is annotated with `@Secured(IS_AUTHENTICATED)`, and you're not yet authenticated:
```

sdelamo commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+image::cloud-oidc-oracle/run3.png[]
+
+After you're redirected to the home page you'll see that you're authenticated and now have a `Logout` link, and several authentication details are displayed:
```
should this be in a common snippet in base?

kroegerk replied:
```suggestion
After you're redirected to the home page, you'll see that you're authenticated and now have a `Logout` link, and several authentication details are displayed:
```

sdelamo commented on guides/micronaut-cloud-oidc-oracle/src/main/resources/application.yml:
```diff
@@ -0,0 +1,20 @@
+micronaut:
```
should `application.yml` be in base?

sdelamo reviewed (changes requested):
We have many OpenID Connect guides? Should we migrate the other guides to use the base in a separate PR?

sdelamo commented:
@kroegerk can you proofread it?

kroegerk commented on guides/micronaut-cloud-oidc-base/start.adoc:
```diff
@@ -0,0 +1,6 @@
+common:header.adoc[]
+
+The application will use OpenID Connect (OIDC), which is an identity layer on top of the OAuth 2.0 protocol, allowing verification of the identity of an end user and obtaining basic profile information.
```
```suggestion
The application will use OpenID Connect (OIDC), which is an identity layer on top of the OAuth 2.0 protocol that verifies the end user's identity and obtains basic profile information.
```

kroegerk commented on guides/micronaut-cloud-oidc-base/start.adoc:
```diff
@@ -0,0 +1,6 @@
+common:header.adoc[]
+
+The application will use OpenID Connect (OIDC), which is an identity layer on top of the OAuth 2.0 protocol, allowing verification of the identity of an end user and obtaining basic profile information.
+Under the hood, the OIDC APIs are easier to work with than OAuth, using JSON instead of XML, but in practice Micronaut Security handles the details for you.
```
```suggestion
Under the hood, the OIDC APIs are easier to work with than OAuth because they use JSON instead of XML, but in practice, Micronaut Security handles the details for you.
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/metadata.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "base": "micronaut-cloud-oidc-base",
+  "title": "Authenticate Micronaut Application users with OpenID Connect and Oracle IDCS",
+  "intro": "Learn how to create a secured Micronaut Application and authenticate with OpenID Connect in Oracle Cloud.",
```
```suggestion
  "intro": "Learn how to create a secured Micronaut application and authenticate with OpenID Connect in Oracle Cloud.",
```

"application" is not a proper noun

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+=== Create a federated user
+
+https://cloud.oracle.com/[Login to your Oracle Cloud tenancy] as an admin (or a user with sufficient permissions to create users and applications), and from the Oracle Cloud Menu select "Identity & Security" and then "Federation":
```
```suggestion
https://cloud.oracle.com/[Log in to your Oracle Cloud tenancy] as an admin (or a user with sufficient permissions to create users and applications), and from the Oracle Cloud Menu, select "Identity & Security" and then "Federation":
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+image::cloud-oidc-oracle/federation05.png[]
+
+Login as the user to verify the password, then log out and back in as the admin user.
```
```suggestion
Log in as the user to verify the password, then log out and back in as the admin user.
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+Click "Configure this application as a client now", and in the `Allowed Grant Types` section check `Client Credentials` and `Authorization Code`.
+
+Check `Allow non-HTTPS URLs` for now to make testing easier and avoid having to use HTTPS locally.
+Be sure to unselect this in production applications which should always use HTTPS.
```
```suggestion
Be sure to unselect this in production applications, which should always use HTTPS.
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+Check `Allow non-HTTPS URLs` for now to make testing easier and avoid having to use HTTPS locally.
+Be sure to unselect this in production applications which should always use HTTPS.
+
+Enter `\http://localhost:8080/oauth/callback/oci` for "Redirect URL", `\http://localhost:8080/logout` for "Logout URL", and `\http://localhost:8080/` for "Post Logout Redirect URL":
```
I think it might be easier for a user to follow these directions if they were split up as so:

```suggestion
Enter the following:

 - Redirect URL: `\http://localhost:8080/oauth/callback/oci` 
 - Logout URL: `\http://localhost:8080/logout`
 - Post Logout Redirect URL: `\http://localhost:8080/`
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+callout:get[number=4,arg0=index,arg1=/]
+callout:secured-is-authenticated[number=5]
+callout:get[number=6,arg0=secured,arg1=/secure]
+<7> This method simply returns a model map that will be rendered   as JSON since there's no `@View` annotation
```
```suggestion
<7> This method simply returns a model map that will be rendered as JSON since there's no `@View` annotation
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+resource:views/home.html[]
+
+Note that we return an empty model in the controller `index()` method, however we are accessing `security` in the Thymeleaf template.
```
```suggestion
Note that we return an empty model in the controller `index()` method. However, we are accessing `security` in the Thymeleaf template.
```

OR

```suggestion
Note that we return an empty model in the controller `index()` method; however, we are accessing `security` in the Thymeleaf template.
```

kroegerk commented on guides/micronaut-cloud-oidc-oracle/micronaut-cloud-oidc-oracle.adoc:
```diff
@@ -0,0 +1,178 @@
+
+image::cloud-oidc-oracle/run4.png[]
+
+Navigate again to http://localhost:8080/secure in a browser and you'll see a simple JSON response:
```
```suggestion
Navigate again to http://localhost:8080/secure in a browser, and you'll see a simple JSON response:
```

graemerocher committed:
Apply suggestions from code review

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

graemerocher committed:
Apply suggestions from code review

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

graemerocher commented:
@sdelamo can you approve?

sdelamo commented:
> @sdelamo can you approve?

I reviewed the PR and I left many comments. I hoped @burtbeckwith could review those comments and address if appropiate.

graemerocher commented:
@burtbeckwith can you address the comments?

burtbeckwith committed:
PR feedback

burtbeckwith commented:
@sdelamo I addressed most of your feedback. I'm going to wait on some of the suggested moves to common base snippets and callouts until the other authors do their cloud-specific guides

burtbeckwith commented:
@sdelamo I wasn't able to get views working in native image (tried , so I removed the TODO I had left. The guide should be ready to merge.

graemerocher commented:
@burtbeckwith what issue did you have with JTE?

burtbeckwith commented:
@graemerocher JTE almost worked, but didn't render the map keys or values in the "User Information" section, so it rendered a list of dashes. I'll create issues in micronaut-views for each variant I tried

sdelamo committed:
twitter card

sdelamo committed:
change publication date to march 31
